content/msc-fearless-extensibility.md

```diff
@@ -0,0 +1,48 @@
+---
+title: "Malleable Systems Challenge: Fearless Extensibility A Conceptual Approach"
+date: "2024-08-17"
+---
+
+This post is a response to the first **[Malleable Systems Challenge Problem](https://forum.malleable.systems/t/a-new-community-activity-challenge-problems/196)**, which explores the theme of **[fearless extensibility](https://forum.malleable.systems/t/challenge-problem-fearless-extensibility/205)**. The challenge asks participants to think deeply about how we can achieve powerful extensibility in systems without introducing significant security and maintenance risks. In this post, I’ll be exploring several dimensions of the problem and proposing ideas for balancing the benefits of extensibility with the inherent challenges. This post is the first in what may become a series, where I draw from real-world examples and systems to illustrate the key concepts.
+
+# Fearless Extensibility: A Conceptual Approach
+
+In this post, we'll explore the challenge of **fearless extensibility** — designing systems that allow users to extend their functionality without introducing undue risks or complexity. We'll examine different dimensions of extensibility and propose ideas for balancing power and safety. Along the way, we'll reference real-world systems and academic research to illustrate key points. <!-- TODO: Make sure I reference real systemd or academic research. >
+
+## Intentional vs. Ad-Hoc Extensibility
+
+One important distinction to make in the design of extensible systems is between **intentional** and **ad-hoc** extensibility.
+
+- **Intentional Extensibility:** Systems are designed with explicit extension points and interfaces that encourage safe, structured modifications. These points are documented and supported by the system's maintainers.
+- **Ad-Hoc Extensibility:** Users extend the system by modifying the existing system, using undocumented APIs, or breaking through other implementation details. These modifications often lead to maintenance headaches, as they are more likely to break with future updates.
```

Comment:
Reference problem example of browser extensions

Reply:
Mention the dual-interface present in class-based object systems where you have both the exposed interface of methods but also the extension interface for children using behavioral inheritance.

Reply:
Give some explanation as to why systems that are intentionally extensible are compositionally more sound, easier to understand, and less brittle.

Probably borrow from the fact that extending your system is in essence a software construction task and you need to be building under some consistent conceptual framework.

Reply:
Explain why intentional extensibility has been necessary: limited interfaces, inability to "upstream", software is not a collaborative project.